### PR TITLE
Fixes for running Ceph, CockroachDB, Minio demos

### LINKED ISSUE #2258: Minio pods in crash loop with `no endpoint pointing to the local machine is found.`
**Is this a bug report or feature request?**
* Bug Report

While testing latest Rook Minio operator from master on Minikube, I see the minio daemon pods are in a crash loop.

**How to reproduce it (minimal and precise):**
```
tests/scripts/minikube.sh up
cd cluster/examples/kubernetes/minio
kubectl create -f operator.yaml
kubectl create -f object-store.yaml
```

Details:
```
> kubectl -n rook-minio get pod
NAME         READY     STATUS             RESTARTS   AGE
my-store-0   0/1       CrashLoopBackOff   5          3m12s
> kubectl -n rook-minio logs my-store-0
Error: Invalid command line arguments server=‘:9000’, args=[http://my-store-0.my-store.rook-minio/data http://my-store-1.my-store.rook-minio/data http://my-store-2.my-store.rook-minio/data http://my-store-3.my-store.rook-minio/data]
```

Further information can be found in this gist: https://gist.github.com/jbw976/23d2a41835caf9f7475d2cab31448e89

### LINKED ISSUE #2264: make codegen broken due to missing generate-groups.sh
**Is this a bug report or feature request?**
* Bug Report

**How to reproduce it (minimal and precise):**
```
make distclean
make vendor
make codegen
```

Behavior:
```
> make codegen
build/codegen/codegen.sh: line 23: ./generate-groups.sh: No such file or directory
make: *** [codegen] Error 1
```

I recently did a `make distclean` to clean out everything it looks like the necessary `k8s.io/code-generator/generate-groups.sh` script used by `make codegen` is not put into place by `make vendor`. 

I believe the recent change to prune unused packages with dep is the cause: https://github.com/rook/rook/commit/eabc9bde1a6db53e2e13382ddf7394b9a6abadab#diff-836546cc53507f6b2d581088903b1785R43

Maybe we need to override this just for the `k8s.io/code-generator` repo.

### PR DESCRIPTION
**Pull request by jbw976**
**Description of your changes:**

- Add cluster domain to minio spec
- Remove port configuration from minio spec
- Include unused packages for codegen repos
- Update Minio version to pick up latest fixes
- Helper to load cockroachdb demo images into minikube

**Which issue is resolved by this Pull Request:**

Resolves #2258 
Resolves #2264 

**Checklist:**
- [x] Documentation has been updated, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

File: tests/scripts/minikube.sh [hunk 1]
```diff
@@ -93,8 +93,9 @@ function check_context() {
 
 function copy_images() {
     if [[ "$1" == "" || "$1" == "ceph" ]]; then
-      echo "copying ceph image"
```
Comment (travisn) [inline tests/scripts/minikube.sh]: should the minikube changes be a separate PR?

File: pkg/operator/minio/controller.go [hunk 2]
```diff
@@ -99,7 +101,7 @@ func (c *MinioController) makeMinioHeadlessService(name, namespace string, spec
 		},
 		Spec: v1.ServiceSpec{
 			Selector:  map[string]string{k8sutil.AppAttr: minioLabel},
-			Ports:     []v1.ServicePort{{Port: spec.Port}},
+			Ports:     []v1.ServicePort{{Port: spec.Network.Ports[0].Port}},
```
Comment (travisn) [inline pkg/operator/minio/controller.go]: what if no ports or more than one port are defined in the crd?

File: pkg/apis/minio.rook.io/v1alpha1/types.go [hunk 3]
```diff
@@ -55,8 +55,8 @@ type ObjectStoreSpec struct {
 	// tolerations).
 	Placement rook.PlacementSpec `json:"placement,omitempty"`
 
-	// Port of operation for the Minio object store.
-	Port int32 `json:"port"`
```
Comment (travisn) [inline pkg/apis/minio.rook.io/v1alpha1/types.go]: are upgrades supported with minio? what happens for existing clusters that have set this port?

Comment (jbw976) [reply pkg/apis/minio.rook.io/v1alpha1/types.go]: upgrades haven't really been supported for minio so far, the types are still in alpha and there is no upgrade guide.  I'd probably take the stance myself that we should still have the freedom to change/evolve the types while they are still in alpha.  I don't think the cost of writing/testing migration logic would really be justified, what are your thoughts? Do you have a strong opinion?

Comment (jbw976) [reply pkg/apis/minio.rook.io/v1alpha1/types.go]: one option would be to have a default port of 9000.  If an existing minio spec doesn't have the new NetworkSpec format, then we could fall back to a default port, so their minio object store spec is still valid?

**Comment by galexrt** [comment]
Are we really sure we need to the cluster domain?

There isn't really a good way to find from the K8S api AFAIK.
The cluster DNS is able to resolve the following names for a StatefulSet (same namespace) see following example.

Headless service `gitlab` for a StatefulSet `gitlab` (namespace: `example-abc`):

* `SERVICENAME` - `gitlab`
* `SERVICENAME.NAMESPACE` - `gitlab.example-abc`
* `SERVICENAME.NAMESPACE.svc` - `gitlab.example-abc.svc` (with cluster domain, but it is not needed as we have `resolv.conf` `search` and `ndots:5` magic)
* `PODNAME.SERVICENAME` - `gitlab-0.gitlab` for a second replica `gitlab-1.gitlab`.
* Only inside each Pod for its own hostname `PODNAME` - In gitlab-0 Pod `gitlab-0` resolves to itself.

So is `PODNAME.SERVICENAME` not working here?
Example for a StatefulSet named `gitlab` (2 replicas):
Pod `gitlab-0`:

```
root@gitlab-0:/home/git/gitlab# host gitlab-0.gitlab
gitlab-0.gitlab.zerbytes-live-proj-sinfra.svc.cluster.local has address 100.64.0.66
root@gitlab-0:/home/git/gitlab# host gitlab-1.gitlab
gitlab-1.gitlab.zerbytes-live-proj-sinfra.svc.cluster.local has address 100.64.3.157
```

For minio here just replace `gitlab` with `minion`

For "performance" optimization (as I'm not 100% sure in which order the "ndots" are applied) adding `NAMESPACE.svc` to the name (e.g., `gitlab-0.gitlab.NAMESPACE.svc`) 
should be "enough".

TL;DR If we don't already use the Service like that, create one headless (`clusterIP: None`) Service for the StatefulSet and a second one for the user as normal.

File: pkg/operator/minio/controller.go (hunk 2 above)
Comment (galexrt) [inline pkg/operator/minio/controller.go]: Is there any (good) reason we want the user to change the port here?
Every (ignoring headless Services for a moment) Service has its own Cluster IP, so where is the point in letting the user choose the port if it doesn't really matter in the end?
The (user) reason "we want to replace an existing S3 but can't change the address without modifying the application code" is not a (good) reason, especially not in a cloud native environment such as Kubernetes.
(Same applies for Pods as they all have their own IP (+ netns))

Comment (jbw976) [reply pkg/operator/minio/controller.go]: That's a good point @galexrt, I don't really know of a strong reason or scenario to be exposing the port that the services use internally.  A client or user would be connecting to the external service they get to define as seen in the example: https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/minio/object-store.yaml#L37-L47

maybe this internal port shouldn't be configurable at all, and we should always just use 9000 for it.

**Comment by jbw976** [comment]
Regarding https://github.com/rook/rook/pull/2266#issuecomment-436146786:
Yes, as described in #2258, Minio fails to start due to invalid arguments if the cluster domain isn't specified in the args we pass to the minio container, because it determines `no endpoint pointing to the local machine is found`.   That error message comes from: https://github.com/minio/minio/blob/master/cmd/endpoint.go#L332

I'm thinking that gets determined here: 
https://github.com/minio/minio/blob/master/cmd/endpoint.go#L147

Inside this function: https://github.com/minio/minio/blob/master/cmd/net.go#L293

But I haven't been able to debug into the minio code more than that.  I can 100% confirm though that without the cluster domain, which was removed in https://github.com/rook/rook/commit/62da01c669809222136fc85e70123518d9a1a0c7, Minio breaks with `no endpoint pointing to the local machine is found`.

**Comment by jbw976** [comment]
@galexrt @travisn i've updated the PR with the changes we discussed in the community meeting yesterday, please take a look!

* Add cluster domain to minio spec, instead of general rook.io network spec
* Remove port configuration from minio spec

**Comment by jbw976** [comment]
@galexrt, I've opened https://github.com/minio/minio/issues/6775 to track in upstream Minio, then noted that in the following comments in our code/docs:
https://github.com/rook/rook/pull/2266/files#diff-7cbf7f2fb6aeac8688be45dc08b54e35R41
https://github.com/rook/rook/pull/2266/files#diff-26867ee278240db250cce54786f88e58R73

I think we should be good here now once I have a green build :)

**Comment by jbw976** [comment]
The last 2 builds have had all green pipelines for AWS, but the 1 GCE pipeline hasn't been started at all:
https://jenkins.rook.io/blue/organizations/jenkins/rook%2Frook/detail/PR-2266/4/pipeline/
https://jenkins.rook.io/blue/organizations/jenkins/rook%2Frook/detail/PR-2266/5/pipeline/

The only hits in the console log for gce or 1.9.x are below:
```
[Pipeline] [gce_1.9.x] { (Branch: gce_1.9.x)
[Pipeline] [gce_1.9.x] node
19:26:49 [gce_1.9.x] Still waiting to schedule task
19:26:49 [gce_1.9.x] All nodes of label ‘gce_1.9.x’ are offline
```
and then much later on (almost 2 hours):
```
Cancelling nested steps due to timeout
[Pipeline] [gce_1.9.x] // node
[Pipeline] [gce_1.9.x] }
21:12:25 [gce_1.9.x] Failed in branch gce_1.9.x
```
It's not clear to me if something is up with GCE, or if a quota is being hit, or something else.
